File: nodejs/routes/route.js
const express = require('express');
const router = express.Router();
const knex = require('../db/connection');

router.get('/', async (req, res) => {
    try {
        const hexes = await knex('route').select(
          'route.*',
          knex.raw('sector_x * 32 + hex_x -1  as origin_x'),
          knex.raw('sector_y * 40 - hex_y -1 as origin_y')
          ).orderBy(['year', 'day']);
        res.status(200).json(hexes);
    } catch (error) {
        console.error(error);
        res.status(500).json({ error: 'Failed to retrieve route' });
    }
});

module.exports = router;

File: ruby/app/models/route.rb
class Route < ApplicationRecord
  self.table_name = "route"

  validates :ship_id, :year, :day, :sector_x, :sector_y, :hex_x, :hex_y, presence: true
  belongs_to :sector, foreign_key: [:sector_x, :sector_y], primary_key: [:x, :y], optional: true
end
